src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -1274,42 +1266,49 @@ export default class Blockchain extends Emittery.Typed<
         Buffer /*codeHash*/
       ])[2];
 
-      let keys: Buffer[] = [];
-      return new Promise((resolve, reject) => {
-        storageTrie
-          .createReadStream()
-          .on("data", data => {
-            keys.push(data.key);
-          })
-          .on("end", () => {
-            // #4 - sort and filter keys
-            const sortedKeys = keys.sort((a, b) => Buffer.compare(a, b));
-
-            // find starting point in array of sorted keys
-            const startKeyBuffer = Data.from(startKey).toBuffer();
-            keys = sortedKeys.filter(key => {
-              if (Buffer.compare(startKeyBuffer, key) <= 0) {
-                return key;
-              }
-            });
+      return new Promise<T>((resolve, reject) => {
+        const startKeyBuffer = Data.from(startKey).toBuffer();
+        const compare = (a: Buffer, b: Buffer) => a.compare(b) < 0;
+
+        const keys: Buffer[] = [];
+        const handleData = ({ key }) => {
+          // ignore anything that comes before our starting point
+          if (startKeyBuffer.compare(key) > 0) return;
+
+          // #4 - sort and filter keys
+          // insert the key exactly where it needs to go in the array
+          const position = findInsertPosition(keys, key, compare);
+          // ignore if the value couldn't possibly be relevant
+          if (position > maxResult) return;
+          keys.splice(position, 0, key);
+        };
 
-            // only take the maximum number of entries requested
-            keys = keys.slice(0, maxResult + 1);
-            if (keys.length > maxResult) {
-              // assign nextKey and remove it from array of keys
-              const nextKey = keys.pop();
-              result.nextKey = Data.from(nextKey).toJSON();
-            }
+        const handleEnd = () => {
+          if (keys.length > maxResult) {
```

Comment:
If we have more data than we want to return we need to trim it down, and grab the `nextKey` as well.